In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
cleaned_data = pd.read_csv("NYPD_Complaint_Data_Historic.csv")

/Users/peimengsui/anaconda/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
cleaned_data.head()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM      RPT_DT  \
0   101109527   12/31/2015     23:45:00          NaN          NaN  12/31/2015   
1   153401121   12/31/2015     23:36:00          NaN          NaN  12/31/2015   
2   569369778   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
3   968417082   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
4   641637920   12/31/2015     23:25:00   12/31/2015     23:30:00  12/31/2015   

   KY_CD                        OFNS_DESC  PD_CD  \
0    113                          FORGERY  729.0   
1    101  MURDER & NON-NEGL. MANSLAUGHTER    NaN   
2    117                  DANGEROUS DRUGS  503.0   
3    344     ASSAULT 3 & RELATED OFFENSES  101.0   
4    344     ASSAULT 3 & RELATED OFFENSES  101.0   

                          PD_DESC              ...               ADDR_PCT_CD  \
0  FORGERY,ETC.,UNCLASSIFIED-FELO              ...                      44.0   
1                             NaN              ...                     103.0   
2  CONTROLLED SUBSTANCE,INTENT TO              ...                      28.0   
3                       ASSAULT 3              ...                     105.0   
4                       ASSAULT 3              ...                      13.0   

  LOC_OF_OCCUR_DESC    PREM_TYP_DESC PARKS_NM  HADEVELOPT X_COORD_CD  \
0            INSIDE   BAR/NIGHT CLUB      NaN         NaN  1007314.0   
1           OUTSIDE              NaN      NaN         NaN  1043991.0   
2               NaN            OTHER      NaN         NaN   999463.0   
3            INSIDE  RESIDENCE-HOUSE      NaN         NaN  1060183.0   
4          FRONT OF            OTHER      NaN         NaN   987606.0   

  Y_COORD_CD   Latitude  Longitude                        Lat_Lon  
0   241257.0  40.828848 -73.916661  (40.828848333, -73.916661142)  
1   193406.0  40.697338 -73.784557  (40.697338138, -73.784556739)  
2   231690.0  40.802607 -73.945052  (40.802606608, -73.945051911)  
3   177862.0  40.654549 -73.726339  (40.654549444, -73.726338791)  
4   208148.0  40.738002 -73.987891     (40.7380024, -73.98789129)  

[5 rows x 24 columns]

In [4]:
cleaned_data.describe()

/Users/peimengsui/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


CMPLNT_NUM         KY_CD         PD_CD   ADDR_PCT_CD    X_COORD_CD  \
count  5.101231e+06  5.101231e+06  5.096657e+06  5.100841e+06  4.913085e+06   
mean   5.499733e+08  2.932191e+02  4.179772e+02  6.346322e+01  1.004610e+06   
std    2.597636e+08  1.480821e+02  2.179312e+02  3.444424e+01  2.144742e+04   
min    1.000002e+08  1.010000e+02  1.010000e+02  1.000000e+00  9.133190e+05   
25%    3.250451e+08  1.170000e+02           NaN           NaN           NaN   
50%    5.499458e+08  3.410000e+02           NaN           NaN           NaN   
75%    7.748865e+08  3.510000e+02           NaN           NaN           NaN   
max    9.999999e+08  8.810000e+02  9.750000e+02  1.230000e+02  1.067298e+06   

         Y_COORD_CD      Latitude     Longitude  
count  4.913085e+06  4.913085e+06  4.913085e+06  
mean   2.068627e+05  4.073443e+01 -7.392651e+01  
std    3.047765e+04  8.365685e-02  7.734681e-02  
min    1.208290e+05  4.049806e+01 -7.425508e+01  
25%             NaN           NaN           NaN  
50%             NaN           NaN           NaN  
75%             NaN           NaN           NaN  
max    2.718200e+05  4.091272e+01 -7.370032e+01

## Get base type for each column

In [5]:
cleaned_data.dtypes

CMPLNT_NUM             int64
CMPLNT_FR_DT          object
CMPLNT_FR_TM          object
CMPLNT_TO_DT          object
CMPLNT_TO_TM          object
RPT_DT                object
KY_CD                  int64
OFNS_DESC             object
PD_CD                float64
PD_DESC               object
CRM_ATPT_CPTD_CD      object
LAW_CAT_CD            object
JURIS_DESC            object
BORO_NM               object
ADDR_PCT_CD          float64
LOC_OF_OCCUR_DESC     object
PREM_TYP_DESC         object
PARKS_NM              object
HADEVELOPT            object
X_COORD_CD           float64
Y_COORD_CD           float64
Latitude             float64
Longitude            float64
Lat_Lon               object
dtype: object

## Check Uniqueness of Complaint Number

In [6]:
len(cleaned_data)==len(cleaned_data.CMPLNT_NUM.unique())

True

## Check if valid date

In [11]:
sum(cleaned_data.CMPLNT_FR_DT.isnull())

655

In [12]:
sum(cleaned_data.CMPLNT_FR_TM.isnull())

48

In [16]:
sum(cleaned_data.CMPLNT_FR_TM.isnull() | cleaned_data.CMPLNT_FR_DT.isnull())

702

In [21]:
sum(cleaned_data.CMPLNT_FR_TM.isnull() & cleaned_data.CMPLNT_FR_DT.isnull())

1

In [17]:
sum(cleaned_data.CMPLNT_TO_DT.isnull())

1391478

In [19]:
sum(cleaned_data.CMPLNT_TO_TM.isnull())

1387785

In [20]:
sum(cleaned_data.CMPLNT_TO_TM.isnull() | cleaned_data.CMPLNT_TO_DT.isnull())

1391993

In [22]:
sum(cleaned_data.CMPLNT_TO_TM.isnull() & cleaned_data.CMPLNT_TO_DT.isnull())

1387270

In [23]:
cleaned_data["CMPLNT_FR_YR"]=pd.to_numeric(cleaned_data.CMPLNT_FR_DT.str[-4:],errors='ignore')
cleaned_data["CMPLNT_TO_YR"]=pd.to_numeric(cleaned_data.CMPLNT_TO_DT.str[-4:],errors='ignore')
cleaned_data["CMPLNT_FR_MO"]=pd.to_numeric(cleaned_data.CMPLNT_FR_DT.str[0:2],errors='ignore')
cleaned_data["CMPLNT_TO_MO"]=pd.to_numeric(cleaned_data.CMPLNT_TO_DT.str[0:2],errors='ignore')
cleaned_data["CMPLNT_FR_DY"]=pd.to_numeric(cleaned_data.CMPLNT_FR_DT.str[3:5],errors='ignore')
cleaned_data["CMPLNT_TO_DY"]=pd.to_numeric(cleaned_data.CMPLNT_TO_DT.str[3:5],errors='ignore')

In [24]:
cleaned_data.describe()

/Users/peimengsui/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


CMPLNT_NUM         KY_CD         PD_CD   ADDR_PCT_CD    X_COORD_CD  \
count  5.101231e+06  5.101231e+06  5.096657e+06  5.100841e+06  4.913085e+06   
mean   5.499733e+08  2.932191e+02  4.179772e+02  6.346322e+01  1.004610e+06   
std    2.597636e+08  1.480821e+02  2.179312e+02  3.444424e+01  2.144742e+04   
min    1.000002e+08  1.010000e+02  1.010000e+02  1.000000e+00  9.133190e+05   
25%    3.250451e+08  1.170000e+02           NaN           NaN           NaN   
50%    5.499458e+08  3.410000e+02           NaN           NaN           NaN   
75%    7.748865e+08  3.510000e+02           NaN           NaN           NaN   
max    9.999999e+08  8.810000e+02  9.750000e+02  1.230000e+02  1.067298e+06   

         Y_COORD_CD      Latitude     Longitude  CMPLNT_FR_YR  CMPLNT_TO_YR  \
count  4.913085e+06  4.913085e+06  4.913085e+06  5.100576e+06  3.709753e+06   
mean   2.068627e+05  4.073443e+01 -7.392651e+01  2.010356e+03  2.010582e+03   
std    3.047765e+04  8.365685e-02  7.734681e-02  3.157654e+00  2.885140e+00   
min    1.208290e+05  4.049806e+01 -7.425508e+01  1.015000e+03  1.912000e+03   
25%             NaN           NaN           NaN           NaN           NaN   
50%             NaN           NaN           NaN           NaN           NaN   
75%             NaN           NaN           NaN           NaN           NaN   
max    2.718200e+05  4.091272e+01 -7.370032e+01  2.015000e+03  2.090000e+03   

       CMPLNT_FR_MO  CMPLNT_TO_MO  CMPLNT_FR_DY  CMPLNT_TO_DY  
count  5.100576e+06  3.709753e+06  5.100576e+06  3.709753e+06  
mean   6.546097e+00  6.587604e+00  1.553795e+01  1.570377e+01  
std    3.381710e+00  3.379293e+00  8.834568e+00  8.799740e+00  
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  
25%             NaN           NaN           NaN           NaN  
50%             NaN           NaN           NaN           NaN  
75%             NaN           NaN           NaN           NaN  
max    1.200000e+01  1.200000e+01  3.100000e+01  3.100000e+01

In [33]:
cleaned_data[cleaned_data["CMPLNT_FR_YR"]<1900]

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
27341    211843983   12/04/1015     16:45:00          NaN          NaN   
27342    131106711   12/04/1015     12:30:00   12/10/2015     12:40:00   
39052    821425869   11/25/1015     14:30:00   12/01/2015     14:40:00   
48548    161924074   09/26/1015     12:11:00   11/23/2015     13:00:00   
72345    414788103   10/27/1015     19:30:00          NaN          NaN   
89523    148685327   10/17/1015     16:00:00   10/23/2015     16:05:00   
119248   704380800   09/16/1015     12:00:00   09/22/2015     12:30:00   

            RPT_DT  KY_CD                     OFNS_DESC  PD_CD  \
27341   12/10/2015    106                FELONY ASSAULT  109.0   
27342   12/10/2015    344  ASSAULT 3 & RELATED OFFENSES  101.0   
39052   12/01/2015    109                 GRAND LARCENY  405.0   
48548   11/23/2015    109                 GRAND LARCENY  415.0   
72345   11/05/2015    341                 PETIT LARCENY  338.0   
89523   10/23/2015    126       MISCELLANEOUS PENAL LAW  198.0   
119248  10/01/2015    233                    SEX CRIMES  175.0   

                                      PD_DESC      ...      Y_COORD_CD  \
27341                ASSAULT 2,1,UNCLASSIFIED      ...        216641.0   
27342                               ASSAULT 3      ...        211854.0   
39052   LARCENY,GRAND BY THEFT OF CREDIT CARD      ...        200707.0   
48548          LARCENY,GRAND FROM PERSON,PICK      ...        232339.0   
72345          LARCENY,PETIT FROM BUILDING,UN      ...        184874.0   
89523                     CRIMINAL CONTEMPT 1      ...        237834.0   
119248                       SEXUAL ABUSE 3,2      ...             NaN   

         Latitude  Longitude                        Lat_Lon  CMPLNT_FR_YR  \
27341   40.761311 -73.973083  (40.761311039, -73.973083376)        1015.0   
27342   40.748137 -73.906245  (40.748136928, -73.906245306)        1015.0   
39052   40.717573 -73.962535  (40.717573206, -73.962534952)        1015.0   
48548   40.804384 -73.937422  (40.804384046, -73.937421669)        1015.0   
72345   40.674073 -73.894438  (40.674072961, -73.894438323)        1015.0   
89523   40.819435 -73.894927  (40.819435407, -73.894926721)        1015.0   
119248        NaN        NaN                            NaN        1015.0   

       CMPLNT_TO_YR CMPLNT_FR_MO CMPLNT_TO_MO CMPLNT_FR_DY  CMPLNT_TO_DY  
27341           NaN         12.0          NaN          4.0           NaN  
27342        2015.0         12.0         12.0          4.0          10.0  
39052        2015.0         11.0         12.0         25.0           1.0  
48548        2015.0          9.0         11.0         26.0          23.0  
72345           NaN         10.0          NaN         27.0           NaN  
89523        2015.0         10.0         10.0         17.0          23.0  
119248       2015.0          9.0          9.0         16.0          22.0  

[7 rows x 30 columns]

In [35]:
cleaned_data[cleaned_data["CMPLNT_TO_YR"]>2020]

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
3363286   624021373   04/06/2009     12:54:00   04/06/2090          NaN   

             RPT_DT  KY_CD      OFNS_DESC  PD_CD  \
3363286  04/06/2009    341  PETIT LARCENY  333.0   

                                PD_DESC      ...      Y_COORD_CD   Latitude  \
3363286  LARCENY,PETIT FROM STORE-SHOPL      ...        166203.0  40.622873   

         Longitude                        Lat_Lon  CMPLNT_FR_YR CMPLNT_TO_YR  \
3363286 -74.004767  (40.622873428, -74.004767265)        2009.0       2090.0   

        CMPLNT_FR_MO CMPLNT_TO_MO CMPLNT_FR_DY  CMPLNT_TO_DY  
3363286          4.0          4.0          6.0           6.0  

[1 rows x 30 columns]

In [ ]:
cleaned_data=data[(data["CMPLNT_FR_YR"]>1900) & (data["CMPLNT_TO_YR"]<2020)

In [48]:
cleaned_data["CMPLNT_FR_DT"]=pd.to_datetime(cleaned_data["CMPLNT_FR_DT"],errors="coerce",format="%m/%d/%Y")
cleaned_data["CMPLNT_TO_DT"]=pd.to_datetime(cleaned_data["CMPLNT_TO_DT"],errors="coerce",format="%m/%d/%Y")

In [49]:
sum(cleaned_data.CMPLNT_FR_DT.isnull())

662

In [50]:
sum(cleaned_data.CMPLNT_TO_DT.isnull())

1391478

In [51]:
len(cleaned_data)

5101231

In [46]:
cleaned_data.CMPLNT_FR_DT=pd.to_datetime(cleaned_data.CMPLNT_FR_DT,errors="coerce",format="%m/%d/%Y")

In [47]:
cleaned_data.dtypes

CMPLNT_NUM                    int64
CMPLNT_FR_DT         datetime64[ns]
CMPLNT_FR_TM                 object
CMPLNT_TO_DT         datetime64[ns]
CMPLNT_TO_TM                 object
RPT_DT               datetime64[ns]
KY_CD                         int64
OFNS_DESC                    object
PD_CD                       float64
PD_DESC                      object
CRM_ATPT_CPTD_CD             object
LAW_CAT_CD                   object
JURIS_DESC                   object
BORO_NM                      object
ADDR_PCT_CD                 float64
LOC_OF_OCCUR_DESC            object
PREM_TYP_DESC                object
PARKS_NM                     object
HADEVELOPT                   object
X_COORD_CD                  float64
Y_COORD_CD                  float64
Latitude                    float64
Longitude                   float64
Lat_Lon                      object
dtype: object

In [40]:
sum(data.CMPLNT_FR_DT.isnull())

655

In [8]:
data.describe()

/Users/peimengsui/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


CMPLNT_NUM         KY_CD         PD_CD   ADDR_PCT_CD    X_COORD_CD  \
count  5.101231e+06  5.101231e+06  5.096657e+06  5.100841e+06  4.913085e+06   
mean   5.499733e+08  2.932191e+02  4.179772e+02  6.346322e+01  1.004610e+06   
std    2.597636e+08  1.480821e+02  2.179312e+02  3.444424e+01  2.144742e+04   
min    1.000002e+08  1.010000e+02  1.010000e+02  1.000000e+00  9.133190e+05   
25%    3.250451e+08  1.170000e+02           NaN           NaN           NaN   
50%    5.499458e+08  3.410000e+02           NaN           NaN           NaN   
75%    7.748865e+08  3.510000e+02           NaN           NaN           NaN   
max    9.999999e+08  8.810000e+02  9.750000e+02  1.230000e+02  1.067298e+06   

         Y_COORD_CD      Latitude     Longitude  CMPLNT_FR_YR  CMPLNT_TO_YR  \
count  4.913085e+06  4.913085e+06  4.913085e+06  5.100576e+06  3.709753e+06   
mean   2.068627e+05  4.073443e+01 -7.392651e+01  2.010356e+03  2.010582e+03   
std    3.047765e+04  8.365685e-02  7.734681e-02  3.157654e+00  2.885140e+00   
min    1.208290e+05  4.049806e+01 -7.425508e+01  1.015000e+03  1.912000e+03   
25%             NaN           NaN           NaN           NaN           NaN   
50%             NaN           NaN           NaN           NaN           NaN   
75%             NaN           NaN           NaN           NaN           NaN   
max    2.718200e+05  4.091272e+01 -7.370032e+01  2.015000e+03  2.090000e+03   

       CMPLNT_FR_MO  CMPLNT_TO_MO  CMPLNT_FR_DY  CMPLNT_TO_DY  
count  5.100576e+06  3.709753e+06  5.100576e+06  3.709753e+06  
mean   6.546097e+00  6.587604e+00  1.553795e+01  1.570377e+01  
std    3.381710e+00  3.379293e+00  8.834568e+00  8.799740e+00  
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  
25%             NaN           NaN           NaN           NaN  
50%             NaN           NaN           NaN           NaN  
75%             NaN           NaN           NaN           NaN  
max    1.200000e+01  1.200000e+01  3.100000e+01  3.100000e+01

### Found some CMPLNT_FR_YR be 1015,and CMPLNT_TO_YR be 2090 might be considered invalid, exclude where less than 

In [9]:
data[data["CMPLNT_TO_YR"]>2020]

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
3363286   624021373   04/06/2009     12:54:00   04/06/2090          NaN   

             RPT_DT  KY_CD      OFNS_DESC  PD_CD  \
3363286  04/06/2009    341  PETIT LARCENY  333.0   

                                PD_DESC      ...      Y_COORD_CD   Latitude  \
3363286  LARCENY,PETIT FROM STORE-SHOPL      ...        166203.0  40.622873   

         Longitude                        Lat_Lon  CMPLNT_FR_YR CMPLNT_TO_YR  \
3363286 -74.004767  (40.622873428, -74.004767265)        2009.0       2090.0   

        CMPLNT_FR_MO CMPLNT_TO_MO CMPLNT_FR_DY  CMPLNT_TO_DY  
3363286          4.0          4.0          6.0           6.0  

[1 rows x 30 columns]

In [10]:
cleaned_data=data[(data["CMPLNT_FR_YR"]>1900) & (data["CMPLNT_TO_YR"]<2020)]

In [ ]:
sum(cleaned_data.CMPLNT_FR_DT.isnull())

## Convert to datetime

In [11]:
cleaned_data["begin_datetime"]=cleaned_data.CMPLNT_FR_DT+" "+cleaned_data.CMPLNT_FR_TM
cleaned_data["end_datetime"]=cleaned_data.CMPLNT_TO_DT+" "+cleaned_data.CMPLNT_TO_TM

/Users/peimengsui/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/Users/peimengsui/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


In [90]:
cleaned_data["begin_datetime"]=pd.to_datetime(cleaned_data.begin_datetime,errors="ignore")
cleaned_data["end_datetime"] = pd.to_datetime(cleaned_data.end_datetime,errors="ignore")

/Users/peimengsui/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/Users/peimengsui/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


In [91]:
cleaned_data.head()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM      RPT_DT  \
4   641637920   12/31/2015     23:25:00   12/31/2015     23:30:00  12/31/2015   
5   365661343   12/31/2015     23:18:00   12/31/2015     23:25:00  12/31/2015   
7   265023856   12/31/2015     23:15:00   12/31/2015     23:15:00  12/31/2015   
8   989238731   12/31/2015     23:15:00   12/31/2015     23:30:00  12/31/2015   
9   415095955   12/31/2015     23:10:00   12/31/2015     23:10:00  12/31/2015   

   KY_CD                     OFNS_DESC  PD_CD                         PD_DESC  \
4    344  ASSAULT 3 & RELATED OFFENSES  101.0                       ASSAULT 3   
5    106                FELONY ASSAULT  109.0        ASSAULT 2,1,UNCLASSIFIED   
7    118             DANGEROUS WEAPONS  792.0        WEAPONS POSSESSION 1 & 2   
8    344  ASSAULT 3 & RELATED OFFENSES  101.0                       ASSAULT 3   
9    341                 PETIT LARCENY  338.0  LARCENY,PETIT FROM BUILDING,UN   

          ...           Longitude                        Lat_Lon CMPLNT_FR_YR  \
4         ...          -73.987891     (40.7380024, -73.98789129)       2015.0   
5         ...          -73.957111  (40.665022689, -73.957110763)       2015.0   
7         ...          -73.910398  (40.845707148, -73.910398033)       2015.0   
8         ...          -73.891900  (40.856711291, -73.891899956)       2015.0   
9         ...          -73.963623   (40.765617688, -73.96362342)       2015.0   

  CMPLNT_TO_YR  CMPLNT_FR_MO CMPLNT_TO_MO CMPLNT_FR_DY CMPLNT_TO_DY  \
4       2015.0          12.0         12.0         31.0         31.0   
5       2015.0          12.0         12.0         31.0         31.0   
7       2015.0          12.0         12.0         31.0         31.0   
8       2015.0          12.0         12.0         31.0         31.0   
9       2015.0          12.0         12.0         31.0         31.0   

        begin_datetime         end_datetime  
4  12/31/2015 23:25:00  12/31/2015 23:30:00  
5  12/31/2015 23:18:00  12/31/2015 23:25:00  
7  12/31/2015 23:15:00  12/31/2015 23:15:00  
8  12/31/2015 23:15:00  12/31/2015 23:30:00  
9  12/31/2015 23:10:00  12/31/2015 23:10:00  

[5 rows x 32 columns]

In [36]:
cleaned_data["duration"]=cleaned_data.end_datetime-cleaned_data.begin_datetime

TypeError: unsupported operand type(s) for -: 'str' and 'str'

## Check Report Date if valid

In [24]:
cleaned_data.RPT_DT = pd.to_datetime(cleaned_data.RPT_DT,errors="ignore",format="%m/%d/%Y")

In [25]:
min(cleaned_data.RPT_DT)

Timestamp('2006-01-01 00:00:00')

In [26]:
max(cleaned_data.RPT_DT)

Timestamp('2015-12-31 00:00:00')

In [27]:
sum(cleaned_data.RPT_DT.isnull())

0

## Check Offense Classification Code

In [28]:
cleaned_data.KY_CD.value_counts()

341    822508
578    604141
344    521611
351    433368
109    429205
235    285801
361    283102
105    198774
107    191407
106    184124
126    112266
110    102061
359    100134
347     73733
236     73680
121     72421
352     66549
117     62683
348     59142
112     56768
118     50563
113     49304
233     44130
340     32057
232     20377
358     17739
360     16064
114     13984
353     13826
104     13791
        ...  
675      1450
342       908
572       829
363       810
346       750
356       618
120       389
230       300
677       239
571       207
237       168
115       122
122       122
234       111
354        95
102        93
455        59
119        48
349        47
366        45
103        33
676        31
672        20
685        18
460        16
881        13
357        11
123         4
362         3
577         3
Name: KY_CD, dtype: int64

In [29]:
sum(cleaned_data.KY_CD.isnull())

0

In [30]:
len(cleaned_data.KY_CD.unique())

74

## Description of offense corresponding with key code

In [31]:
len(cleaned_data.OFNS_DESC.unique())

71

In [32]:
sum(cleaned_data.OFNS_DESC.isnull())

18840

## Three digit internal classification code

In [33]:
cleaned_data.PD_CD.value_counts()

101.0    438130
638.0    368119
639.0    281278
637.0    235959
338.0    221734
333.0    215446
254.0    188807
321.0    178035
567.0    173044
109.0    154046
259.0    146449
438.0    121739
511.0     99943
397.0     99645
441.0     91883
258.0     89416
221.0     82516
339.0     76594
782.0     73653
905.0     73005
113.0     70185
205.0     60001
739.0     54563
198.0     54032
419.0     49615
421.0     48550
405.0     42892
729.0     41503
916.0     41455
793.0     41046
          ...  
396.0         3
685.0         3
689.0         3
622.0         3
646.0         3
796.0         3
589.0         3
845.0         3
820.0         2
614.0         2
836.0         2
861.0         2
787.0         2
387.0         2
846.0         2
860.0         2
523.0         2
694.0         2
184.0         2
178.0         2
839.0         1
522.0         1
678.0         1
815.0         1
814.0         1
819.0         1
885.0         1
146.0         1
148.0         1
975.0         1
Name: PD_CD, dtype: int6

In [34]:
len(cleaned_data.PD_CD.unique())

416

## Description of internal classification corresponding with PD code

In [35]:
len(cleaned_data.PD_DESC.unique())

404

## Crime Indicator

In [36]:
cleaned_data.CRM_ATPT_CPTD_CD.value_counts()

COMPLETED    5013311
ATTEMPTED      87913
Name: CRM_ATPT_CPTD_CD, dtype: int64

In [37]:
sum(cleaned_data.CRM_ATPT_CPTD_CD.isnull())

7

In [38]:
cleaned_data[cleaned_data.CRM_ATPT_CPTD_CD.isnull()].CRM_ATPT_CPTD_CD

899205     NaN
1325328    NaN
2075706    NaN
2152660    NaN
4746958    NaN
4747214    NaN
4750394    NaN
Name: CRM_ATPT_CPTD_CD, dtype: object

## Level of offense

In [39]:
cleaned_data.LAW_CAT_CD.value_counts()

MISDEMEANOR    2918574
FELONY         1567423
VIOLATION       615234
Name: LAW_CAT_CD, dtype: int64

In [40]:
sum(cleaned_data.LAW_CAT_CD.isnull())

0

## Jurisdiction responsibility description

In [43]:
cleaned_data.JURIS_DESC.value_counts()

N.Y. POLICE DEPT                       4538344
N.Y. HOUSING POLICE                     390853
N.Y. TRANSIT POLICE                     108817
PORT AUTHORITY                           24657
OTHER                                    13575
POLICE DEPT NYC                           8986
DEPT OF CORRECTIONS                       4825
TRI-BORO BRDG TUNNL                       4633
HEALTH & HOSP CORP                        2590
N.Y. STATE POLICE                         1209
METRO NORTH                                531
FIRE DEPT (FIRE MARSHAL)                   514
LONG ISLAND RAILRD                         439
STATN IS RAPID TRANS                       306
N.Y. STATE PARKS                           272
U.S. PARK POLICE                           185
AMTRACK                                    153
NEW YORK CITY SHERIFF OFFICE               134
NYS DEPT TAX AND FINANCE                    77
NYC PARKS                                   71
SEA GATE POLICE DEPT                        30
NYC DEPT ENVI

In [44]:
sum(cleaned_data.JURIS_DESC.isnull())

0